synapse/lib/cell.py

```diff
@@ -4031,3 +4047,219 @@ async def _runReloadFunc(self, name, func):
+        lkey = user + b'cell:apikey' + iden.encode('utf-8')
+        buf = self.slab.get(lkey, db=self.usermetadb)  # None here would be a nexus state inconsistency
+        kdef = s_msgpack.un(buf)  # This includes the shadow key
+        return kdef
+
+    async def listUserApiKeys(self, user=None):
+        # DISCUSS: This could easily be a generator too
+        vals = []
+        lkey = b''
+        if user:
+            lkey = user.encode('utf-8') + b'cell:apikey'
+        for lkey, valu in self.slab.scanByPref(lkey, db=self.usermetadb):
+            useriden, suffix = lkey[:32], lkey[32:]
+            if suffix.startswith(b'cell:apikey'):
+                kdef = s_msgpack.un(valu)
+                kdef.pop('shadow')
+                vals.append((useriden.decode('utf-8'), kdef))
+        return vals
+
+    async def checkUserApiKey(self, key):
+        isok, valu = s_passwd.parseApiKey(key)  # TODO Timecache this API call
+        if isok is False:
+            return False, {'mesg': 'API Key is malformed.'}
+
+        iden, secv = valu
+
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            return False, {'mesg': f'API Key does not exist: {iden}'}
+
+        user = kdef.get('user')
+        udef = await self.getUserDef(user)
+        if udef is None:
+            return False, {'mesg': f'User does not exist for API Key: {iden}', 'user': user}
+
+        if udef.get('locked'):
+            return False, {'mesg': f'User associated with API Key is locked: {iden}',
+                           'user': user, 'name': udef.get('name')}
+
+        duration = kdef.get('duration')
+        if duration is not None:
+            now = s_common.now()
+            if now >= kdef.get('expref') + kdef.get('duration'):
+                return False, {'mesg': f'API Key is expired: {iden}',
+                               'user': user, 'name': udef.get('name')}
+
+        shadow = kdef.pop('shadow')
+        valid = await s_passwd.checkShadowV2(secv, shadow)  # TODO Timecache this API call
+        if valid is False:
+            return False, {'mesg': f'API Key shadow mismatch: {iden}',
+                           'user': user, 'name': udef.get('name')}
+
+        return True, {'kdef': kdef, 'udef': udef}
+
+    async def modUserApiKey(self, iden, key, valu):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist, cannot modify it: {iden}', iden=iden)
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+
+        vals = {
+            'modified': s_common.now()
+        }
+        if key == 'name':
+            vals['name'] = valu
+        else:
+            raise s_exc.BadArg(mesg=f'Cannot set {key} on user API Keys.', name=key)
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
```

Comment:
I think `user:apikey:add` and `user:apikey:update` could be unified into a single `user:apikey:set` since they are required to be idempotent anyway

Reply:
I kept them separate to keep small operations like changing a name ( plus any future keys we may wanted to add as properties) require storing the full blob.

Reply:
Stored in the DB but doesn't have to be stored in the nexus each time; this should be similar to the recent cron changes.

Reply:
Ah, gotcha :+1: To satisfy my OCD could we make it `user:apikey:edit` for consistency? :D

Reply:
( or at least to not add to the inconsistency :rofl: )